src/Bicep.Core/Emit/SourceMapHelper.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.Workspaces;
+using Microsoft.WindowsAzure.ResourceStack.Common.Extensions;
+
+namespace Bicep.Core.Emit
+{
+    public static class SourceMapHelper
+    {
+        public static void AddMapping(
+            IDictionary<string, IDictionary<int, IList<(int start, int end)>>> rawSourceMap,
+            BicepFile bicepFile,
+            SyntaxBase bicepSyntax,
+            PositionTrackingJsonTextWriter jsonWriter,
+            int jsonStartPos)
+        {
+            if (rawSourceMap is null)
+            {
+                throw new ArgumentNullException(nameof(rawSourceMap));
+            }
+
+            var bicepFilePath = bicepFile.FileUri.AbsolutePath;
+            (int bicepLine, _) = TextCoordinateConverter.GetPosition(bicepFile.LineStarts, bicepSyntax.GetPosition());
+
+            // account for leading nodes (decorators)
+            if (bicepSyntax is StatementSyntax syntax)
+            {
+                bicepLine += syntax.LeadingNodes.Count(node => node is Token token && token.Type is TokenType.NewLine);
+            }
+
+            // increment start position if starting on a comma (happens when outputting successive items in objects and arrays)
+            if (jsonWriter.CommaPositions.Contains(jsonStartPos))
+            {
+                jsonStartPos++;
+            }
+
+            var jsonEndPos = jsonWriter.CurrentPos - 1;
+
+            rawSourceMap.AddMapping(
+                bicepFilePath,
+                bicepLine,
+                jsonStartPos,
+                jsonEndPos);
+        }
+
+        public static void AddNestedSourceMap(
+            this IDictionary<string, IDictionary<int, IList<(int start, int end)>>> parentSourceMap,
```

Comment:
Could we have a typed definition for this instead? Using `record` makes this super simple - see https://github.com/Azure/bicep/blob/62f4719d76e408babb4f041a4e7b698ac44b1176/src/Bicep.Core/TypeSystem/ResourceType.cs#L8-L11 for example.

Reply:
Do you mean just "wrapping" this dictionary in a record type for a cleaner look for its typing? e.g.:

`public record RawSourceMap(IDictionary<int, (string, int)>? map)`


`public void SomeMethod(RawSourceMap rawSourceMap) { doSomething(rawSourceMap.map); }`